src/helpers/buffers/send.rs

```diff
@@ -49,35 +57,59 @@ impl SendBuffer {
         }
     }
 
-    pub fn push(&mut self, channel_id: &ChannelId, msg: &MessageEnvelope) -> Option<Vec<u8>> {
+    pub fn push(
+        &mut self,
+        channel_id: &ChannelId,
+        msg: &MessageEnvelope,
+        total_records: NonZeroUsize,
+    ) -> Option<Vec<u8>> {
         debug_assert!(
             msg.payload.len() <= ByteBuf::ELEMENT_SIZE_BYTES,
             "Message payload exceeds the maximum allowed size"
         );
 
-        let buf = if let Some(buf) = self.inner.get_mut(channel_id) {
-            buf
+        let channel = if let Some(channel) = self.inner.get_mut(channel_id) {
+            channel
         } else {
             self.inner.entry(channel_id.clone()).or_insert_with(|| {
+                tracing::trace!("create tx channel {:?}", channel_id);
                 let size =
                     NonZeroUsize::new(self.batch_count.get() * self.items_in_batch.get()).unwrap();
-                FixedSizeByteVec::new(size)
+                SendChannel {
+                    total_records: TotalRecords::Specified(total_records),
+                    sent_records: 0,
+                    buf: FixedSizeByteVec::new(size),
+                }
             })
         };
 
         // TODO: avoid the copy here and size the element size to the message type.
         let mut payload = [0; ByteBuf::ELEMENT_SIZE_BYTES];
         payload[..msg.payload.len()].copy_from_slice(&msg.payload);
-        buf.insert(channel_id, usize::from(msg.record_id), &payload);
-        buf.take(self.items_in_batch.get())
+        channel.buf.insert(channel_id, usize::from(msg.record_id), &payload);
+        let res = channel.buf.take(self.items_in_batch.get());
+        if let Some(ref vec) = res {
+            assert!(vec.len() % ByteBuf::ELEMENT_SIZE_BYTES == 0);
+            channel.sent_records += vec.len() / ByteBuf::ELEMENT_SIZE_BYTES;
+        }
+        if NonZeroUsize::new(channel.sent_records) == Some(channel.total_records.try_into().unwrap()) {
+            assert!(NonZeroUsize::new(channel.buf.taken()) == Some(channel.total_records.try_into().unwrap()));
+            assert!(channel.buf.is_empty());
+            self.inner.remove(&channel_id);
```

Comment:
I am wondering if @martinthomson and you had a chance to discuss whether we need to keep a tombstone after closing the channel. It will cost us something but also may catch some gnarly bugs in the future. Sending through a tombstone channel is an immediate panic. It is more work so even if we decide to do it, it does not need to be in this change, but I am just curious what you think about it

Reply:
I think that if we move to a channel per step, we'll be in a better position to handle that.